core/src/main/clojure/xtdb/operator/scan.clj

```diff
@@ -361,29 +362,68 @@
         (vr/<-root vsr))))
 
   (test-metadata [_mpg]
-    (.test page-idx-pred page-idx)))
+    (.test page-idx-pred page-idx))
+
+  (temporal-bounds [_mpg] (.temporalBounds table-metadata (int page-idx))))
+
+(def ^:private non-constraint-bounds (TemporalBounds.))
 
 (defrecord LiveMergePlanPage [^RelationReader live-rel trie ^LiveHashTrie$Leaf leaf]
   MergePlanPage
   (load-page [_mpg _buffer-pool _vsr-cache]
     (.select live-rel (.mergeSort leaf trie)))
 
-  (test-metadata [_mpg] true))
-
-(defn ->merge-task [mp-pages]
-  (let [leaves (ArrayList.)]
-    (loop [[mp-page & more-mp-pages] mp-pages
-           node-taken? false]
-      (if mp-page
-        (let [take-node? (test-metadata mp-page)]
-
-          (when (or take-node? node-taken?)
-            (.add leaves mp-page))
-
-          (recur more-mp-pages (or node-taken? take-node?)))
-
-        (when node-taken?
-          (vec leaves))))))
+  (test-metadata [_mpg] true)
+
+  (temporal-bounds [_msg] non-constraint-bounds))
+
+(defn max-valid-to [^TemporalBounds tb]
+  (.getUpper (.getValidTo tb)))
+
+(defn min-valid-from [^TemporalBounds tb]
+  (.getLower (.getValidFrom tb)))
+
+(defn ->merge-task
+  ([mp-pages] (->merge-task mp-pages (TemporalBounds.)))
+  ([mp-pages ^TemporalBounds query-bounds]
+   (let [leaves (ArrayList.)]
+     (loop [[mp-page & more-mp-pages] mp-pages
+            node-taken? false
+            largest-valid-to Long/MIN_VALUE
+            non-taken-pages []]
+       (if mp-page
+         (let [^TemporalBounds page-bounds (temporal-bounds mp-page)
+               take-node? (and (.intersects page-bounds query-bounds)
+                               (test-metadata mp-page))]
+
+           (if take-node?
+             (do
+               (.add leaves mp-page)
+               (recur more-mp-pages
+                      true
+                      (cond-> largest-valid-to
+                        (not (instance? LiveMergePlanPage mp-page))
```

Comment:
The instance check was something we wanted to avoid with the `MergePlanNode` protocol. The problem is that on the one hand we don't have metadata on the live-trie and hence are just returning temporal-bounds that are unconstrained (i.e. always take the live trie), on the other hand we don't want these temporal bounds to be included into the max-valid-to calculation.

Reply:
would be good to understand this one a bit better :sweat_smile:

edit: maybe this is a sign that re-using the temporal bounds isn't quite sound?